src/lib/react-query/documents.ts

```diff
@@ -168,3 +193,109 @@ export function documentByVersionIdQueryOptions<D extends DocumentType>({
 		},
 	})
 }
+
+export function createDocumentMutationOptions<D extends WriteableDocumentType>({
+	docType,
+	projectApi,
+	projectId,
+	queryClient,
+}: {
+	docType: D
+	projectApi: MapeoProjectApi
+	projectId: string
+	queryClient: QueryClient
+}) {
+	return {
+		...baseMutationOptions(),
+		mutationFn: async ({
+			value,
+		}): Promise<WriteableDocument<D> & { forks: Array<string> }> => {
+			// @ts-expect-error TS not handling this well
+			return projectApi[docType].create({
+				...value,
+				schemaName: docType,
+			})
+		},
+		onSuccess: () => {
+			queryClient.invalidateQueries({
+				queryKey: getDocumentsQueryKey({
+					projectId,
+					docType,
+				}),
+			})
+		},
+	} satisfies UseMutationOptions<
+		WriteableDocument<D> & { forks: Array<string> },
+		Error,
+		{ value: Omit<WriteableValue<D>, 'schemaName'> }
+	>
+}
+
+export function updateDocumentMutationOptions<D extends WriteableDocumentType>({
+	docType,
+	projectApi,
+	projectId,
+	queryClient,
+}: {
+	docType: D
+	projectApi: MapeoProjectApi
+	projectId: string
+	queryClient: QueryClient
+}) {
+	return {
+		...baseMutationOptions(),
+		mutationFn: async ({
+			versionId,
+			value,
+		}): Promise<WriteableDocument<D> & { forks: Array<string> }> => {
+			// @ts-expect-error TS not handling this well
+			return projectApi[docType].update(versionId, value)
+		},
+		onSuccess: () => {
+			queryClient.invalidateQueries({
+				queryKey: getDocumentsQueryKey({
+					projectId,
+					docType,
+				}),
+			})
+		},
+	} satisfies UseMutationOptions<
+		WriteableDocument<D> & { forks: Array<string> },
+		Error,
+		{ versionId: string; value: Omit<WriteableValue<D>, 'schemaName'> }
+	>
+}
+
+export function deleteDocumentMutationOptions<D extends WriteableDocumentType>({
+	docType,
+	projectApi,
+	projectId,
+	queryClient,
+}: {
+	docType: D
+	projectApi: MapeoProjectApi
+	projectId: string
+	queryClient: QueryClient
+}) {
+	return {
+		...baseMutationOptions(),
+		mutationFn: async ({
+			docId,
+		}): Promise<WriteableDocument<D> & { forks: Array<string> }> => {
+			// @ts-expect-error TS not handling this well
+			return projectApi[docType].delete(docId)
+		},
+		onSuccess: () => {
+			queryClient.invalidateQueries({
+				queryKey: getDocumentsQueryKey({
+					projectId,
+					docType,
+				}),
+			})
```

Comment:
wondering if this should be more granular about what it invalidates. right now this invalidates _all_ queries related to the document type for the relevant project, which is potentially invalidating more than necessary. it might make sense to only invalidate the queries as following:

1. Invalidate the 'get many' query _without cascading_
2. Invalidate the 'get by version id' query _with cascading_
3. Invalidate the 'get by doc id' query _with cascading_

I think this would be an optimization, but maybe it's fine to ignore for now. How we currently have it is "safer" despite potentially doing more work than necessary.